src/lib/utils.ts

```diff
@@ -105,3 +105,9 @@ export const deepUpsertArray = <T>(
   if (!level1) level1 = data[level1Key] = {}
   level1[level2Key] = union(level1[level2Key] ?? [], [value])
 }
+
```

Comment:
Handy little convenience function - just give it your type guard function and an assertion message and you get back a type guard assertion function.